projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+package main
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/csv"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// define a type of struct
+// (!) ` ` are struct tags, it provides metadata which can be used by Go packages like encoding/json, otherwise the unmarshal will use the field names
+type Player struct {
+	Name      string `json:"name"`
+	HighScore int32  `json:"high_score"`
+}
+
+// [1] JSON
+func parseJSON(file []byte, dataSlice *[]Player) error {
+	// try to "unmarshal" the data from the json file into the jsonDataSlice variable
+	err := json.Unmarshal(file, &dataSlice)
+
+	// if we fail to "unmarshal" the json file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	return nil
+}
+
+// [2] Repeated JSON
+func parseRepeatedJSON(file []byte, dataSlice *[]Player) error {
+	// convert the file to a string and then split it on a newline character
+	lines := strings.Split(string(file), "\n")
+
+	// if no lines were parsed then error
+	if len(lines) < 1 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	// initialise a player struct to iteratively store each valid player
```

Comment:
Can you talk through when the value in the `player` variable is copied in this function?

It's interesting to re-use the same `Player` instance for all of the loop iterations, rather than making a new one for each loop iteration - what are the trade-offs here?

Reply:
I re-used the same `player` struct to use the same memory chunk
I don't know what you mean by copied.. do you mean when the Unmarshalling happens or when the appending happens?
If we are constantly overwriting to the same struct there could be some weird behaviour(??)